---
title: "HCATonsilData"
author: "Ramon Massoni-Badosa, Federico Marini, Alan O'Callaghan & Helena Crowell"
date: "`r BiocStyle::doc_date()`"
output: 
  BiocStyle::html_document:
    toc_float: true
vignette: >
  %\VignetteIndexEntry{HCATonsilData}
  %\VignetteEncoding{UTF-8}
  %\VignettePackage{HCATonsilData}
  %\VignetteKeywords{GeneExpression, RNASeq, SingleCell, ATACSeq, Spatial, CITESeq}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: inline
bibliography: HCATonsilData.bib
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  eval = FALSE
)
```


# Introduction

Palatine tonsils are under constant exposure to antigens via the upper
respiratory tract, which makes them a compelling model secondary lymphoid organ
(SLO) to study the interplay between innate and adaptive immune cells
[@ruddle2009secondary]. Tonsils have a non-keratinizing stratified squamous
epithelium, organized into tubular, branched crypts that enlarge the tonsillar
surface. Within the crypts, microfold cells (or M cells) sample antigens at
their apical membrane. Subsequently, antigen presenting cells (APC), such as
dendritic cells (DC), process and present antigens to T cells in the
interfollicular or T cell zone. Alternatively, antigens are kept intact by
follicular dendritic cells (FDC) in lymphoid follicles, where they are
recognized by B cells [@nave2001morphology]. Such recognition triggers the
GC reaction, whereby naive B cells (NBC) undergo clonal selection,
proliferation, somatic hypermutation, class switch recombination (CSR) and
differentiation into long-lived plasma cells (PC) or memory B cells (MBC)
[@de2015dynamics].

In the context of the Human Cell Atlas (HCA) [@regev2018human], we have created 
a taxonomy of 121 cell types and states in a human tonsil. Since the transcritpome
is just a snapshot of a cell's state [@wagner2016revealing], we have added other
layers to define cell identity: single-cell resolved open chromatin epigenomic
landscapes (scATAC-seq and scRNA/ATAC-seq; i.e. Multiome) as well as protein
(CITE-seq), adaptive repertoire (single-cell B and T cell receptor sequencing;
i.e. scBCR-seq and scTCR-seq) and spatial transcriptomics (ST) profiles.

The HCATonsilData package aims to provide programmatic and modular access to the
datasets of the different modalities and cell types of the tonsil atlas.
HCATonsilData also documents how the dataset was generated and archived in
different repositories, from raw fastq files to processed datasets. It also
explains in detail the cell- and sample-level metadata, and allows users to
traceback the annotation of each cell type through an interactive Glossary.


# Getting Started

Load the packages in R

```{r message=FALSE}
library(HCATonsilData)
library(SingleCellExperiment)
library(ExperimentHub)
library(ggplot2)
library(knitr)
library(kableExtra)
```


# Overview of the dataset

We obtained a total of 17 human tonsils, which covered three age groups:
children (n=6, 3-5 years), young adults (n=8, 19-35 years) and old adults
(n=3, 56-65 years). We collected them in a discovery cohort (n=10), which we
used to cluster and annotate cell types; and a validation cohort (n=7), 
which we used to validate the presence, annotation and markers of the discovered
cell types. The following table corresponds to the donor-level metadata:

```{r}
kable(
  donor_metadata,
  format = "markdown",
  caption = "Donor Metadata",
  align = "c"
) %>% kable_styling(full_width = FALSE)
```




Here's a more detailed explanation of the different
types of assay:

* scRNA-seq ([10X Chromium v3](https://www.10xgenomics.com/products/single-cell-gene-expression))
* scATAC-seq ([10X Chromium](https://www.10xgenomics.com/products/single-cell-atac))
* [10X Multiome](https://www.10xgenomics.com/products/single-cell-multiome-atac-plus-gene-expression): joint RNA and ATAC for each cell.
* [CITE-seq](https://www.nature.com/articles/nmeth.4380): joint transcriptome + ~200 protein surface markers for each cell.
* [Spatial transcriptomics]([10X Visium](https://www.10xgenomics.com/products/spatial-gene-expression))

The following heatmap informs about which samples where sequenced with which
technology:

```{r, echo=FALSE, out.width = "100%"}
knitr::include_graphics("tonsil_atlas_cohort.png")
```


The following UMAPs inform about the main cell populations and number of cells
per assay:

```{r, echo=FALSE, out.width = "100%"}
knitr::include_graphics("tonsil_atlas_umaps.png")
```


The dataset was originally analyzed with [Seurat](https://satijalab.org/seurat/).
[@hao2021integrated] Important links:

* [Github repository with the full set of Rmd notebooks needed to go from raw expression matrices to the final objects](TODO add link).
* [Zenodo repository where we deposited the Seurat objects](https://zenodo.org/record/6340174/#.Yl7BH39BxH4).


# Assay types

HCATonsilData provides access to 4 main types of assays: RNA, ATAC, CITE-seq and
Spatial.

## RNA

We can obtain the `SingleCellExperiment` object with transcriptomics (RNA)
data as follows:

```{r}
(sce <- HCATonsilData(assayType = "RNA", cellType = "All"))
table(sce$assay)
```


As you can see, this object consists of 209786 profiled with scRNA-seq (3P)
and 53513 cells profiled with multiome. Initially, we aimed to also use CITE-seq
to cluster cells in our tonsil atlas. However, we noticed that with CITE-seq we
detected 2.75X and 2.98X fewer genes than with scRNA-seq and Multiome, respectively:

```{r eval=FALSE, echo=FALSE}
df <- colData(sce)[, c("assay", "nFeature_RNA")]
df <- as.data.frame(df)
cite <- readRDS("inst/scripts/cite_raw_data/20220215_tonsil_atlas_cite_seurat_obj.rds")
cite <- cite@meta.data[, c("assay", "nFeature_RNA")]
df <- rbind(df, cite)
df$assay <- factor(df$assay, levels = c("3P", "multiome", "CITE-seq"))
levels(df$assay) <- c("scRNA-seq", "multiome", "CITE-seq")
colors <- c("#d52f54", "#4bafdd", "#f0b635")
names(colors) <- levels(df$assay)
ggplot(df, aes(assay, nFeature_RNA, fill = assay)) +
  geom_violin() +
  stat_summary(fun = "mean", geom = "point", color = "black") +
  labs(x = "", y = "Number of detected genes") +
  scale_fill_manual(values = colors, breaks = names(colors)) +
  theme_classic() +
  theme(legend.position = "none")
```

```{r, echo=FALSE, out.width = "100%"}
knitr::include_graphics("tonsil_atlas_n_detected_genes.png")
```

Because we did not want to bias the clustering towards the modality that
provides less information, we first focused on scRNA-seq and Multiome together,
and use CITE-seq to validate the annotation (see below). However, a recent
benchmarking effort showed that scRNA-seq and single-nuclei RNA-seq (such as
Multiome) mix very poorly [@mereu2020benchmarking]. Indeed, we observed massive batch effects 
between both modalities (see manuscript). To overcome it, we found highly variable
genes (HVG) for each modality independently. We then intersected both sets of
HVG to remove modality-specific variation. Following principal component
analysis (PCA), we integrated scRNA-seq and Multiome with Harmony (TODO cite). Harmony
Harmony scales well to atlas-level dataset and ranks among the best-performing
tools in different benchmarks [@korsunsky2019fast]. We have automatized this two-step
approach in [SLOcatoR](https://github.com/massonix/SLOcatoR), an R package that annotates
transcriptomes and chromatin accessibility profiles from SLO. As shown in the
manuscript, we successfully corrected for batch effects while preserving
biological variation.
To cluster cells we followed a top-down approach, from general to specific, 
stratified by levels. In each level, we perform the two-step process describe
above, which finds finer-grained heterogeneity in a recursive manner. Then, we
used the batch-corrected principal components (harmony components), to calculate
a K-nearest neighbor graph (KNN graph). Louvain clustering uses this KNN graph
to assign a cluster to each cell. In level 1, we identified 11 main cell
populations, which we annotated using well-defined marker genes:

```{r}
df <- as.data.frame(colData(sce))
cols <- c("#5A5156", "#E4E1E3", "#F6222E", "#FE00FA", "#16FF32", "#3283FE",
          "#FEAF16", "#B00068","#1CFFCE",  "#90AD1C", "#2ED9FF") 
ggplot(df, aes(UMAP_1_level_1, UMAP_2_level_1, col = annotation_level_1)) +
  geom_point(shape = ".") +
  scale_color_manual(values = cols) +
  labs(title = "Annotation Level 1", x = "UMAP1", y = "UMAP2") +
  theme_classic() +
  theme(
    plot.title = element_text(hjust = 0.5),
    legend.title = element_blank(),
    axis.text = element_blank(),
    axis.ticks = element_blank()
  ) +
  guides(colour = guide_legend(override.aes = list(size = 2, shape = 16)))
```


For each cluster at level 1, we followed multiple levels of integration,
clustering, and removal of poor-quality cells. Finally, we annotated them to
specific cell types and states, which are stored in the variable
`annotation_20220215`.

We can dowload a `SingleCellExperiment` object specific to each of the main
subpopulations defined at level 1 as follows:

```{r}
listCellTypes(assayType = "RNA")
(epithelial <- HCATonsilData(assayType = "RNA", cellType = "epithelial"))
df <- as.data.frame(colData(epithelial))
ggplot(df, aes(UMAP_1_20220215, UMAP_2_20220215, col = annotation_20220215)) +
  geom_point() +
  scale_color_manual(values = cols) +
  labs(title = "Annotation 20220215", x = "UMAP1", y = "UMAP2") +
  theme_classic() +
  theme(
    plot.title = element_text(hjust = 0.5),
    legend.title = element_blank(),
    axis.text = element_blank(),
    axis.ticks = element_blank()
  )
```

As we can see, the annotation we are currently using has a date associated with
it. Initially, we annotated each clusters with the best information we had at
a given point in time, based on previously published literature and to the best
of our knowledge. However, annotations are imperfect by nature, and they need
to be reassessed after new evidence is available or other experts look at the
data. In this setting, we will add a new column to the `colData` slot of each
SingleCellExperiment object every time we update the annotation, with the form
`annotation_date`.
In addition, future versions of HCATonsilData will contain dataframes that
allow to traceback the reasoning for each annotation: which markers where used,
which are the key references, brief explanation of the rationale, etc (TODO).

Here's a brief explanation of all the variables in the colData slot of the
`SingleCellExperiment` objects:


* barcode: the cell barcode. Combination of [GEM well](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/glossary) and cellranger 10X barcode.
* donor_id
* gem_id: we gave a unique hashtag to each GEM well (10X Chip Channel) in our dataset. This allows to traceback all the metadata for a given cell.       
* library_name: each GEM well can give rise to multiple Illumina libraries (example: multiome, cell hashing, etc.).
* assay: 3P (scRNA-seq) or multiome
* sex
* age
* age_group: kid, young adult, old adult
* hospital: hospital where the tonsils where obtained [Hospital Clinic](https://www.clinicbarcelona.org/) or [CIMA](https://cima.cun.es/).
* UMAP_1_level_1, UMAP_2_level_1: UMAP1 coordinates of the figure 1b of our manuscript
* nCount_RNA, nFeature_RNA: number of counts and features (genes) detected per cell.
* pct_mt, pct_ribosomal: percentage of counts that map to mitochondrial (^MT) or ribosomal (^RPS) genes.
* is_hashed: whether we used [cell hashing](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-018-1603-1) or not.
* pDNN_hashing, pDNN_scrublet, pDNN_union: proportion of doublet nearest neighbors (pDNN) using different doublet annotations.
* scrublet_doublet_scores
* scrublet_predicted_doublet
* S.Score, G2M.Score Phase, CC.Difference: outputs of the [CellCycleScoring](https://rdrr.io/cran/Seurat/man/CellCycleScoring.html) from `Seurat`.
* annotation_level_1: annotation used at our level 1 clustering (see above).
* annotation_figure_1: annotation used in the figure 1 of our manuscript. This annotation consisted of grouping the final subtypes into main cell types that were distinguishable in the UMAP.
* annotation_20220215: see above.
* UMAP_1_20220215, UMAP_2_20220215: UMAP coordinates used in figures of the preprint for each cell type.


## ATAC

Coming soon! (TODO)

## CITE-seq

Coming soon! (TODO) For now you can download the original `Seurat` object with the
following code:

```{r}
#code to download specific CITE seq from zenodo
```


## Spatial

Coming soon! (TODO) For now you can download the original `Seurat` object with the
following code:

```{r}
#code to download specific spatialfrom zenodo
```



# Explore data with iSEE

Coming soon! (TODO)

# Exploring information on the Tonsil Data Atlas

Yadda yadda TODO, one can read up info on this in two different manners

```{r eval = TRUE}
library("HCATonsilData")

head(TonsilData_glossary())

TonsilData_cellinfo("Th17")
```


```{r eval = TRUE}
htmltools::includeHTML(
  TonsilData_cellinfo_html("Th17", display_plot = TRUE)
)
```


# Session information

```{r}
sessionInfo()
```


# References
